internal/rpc/digest_md5.go

```diff
@@ -0,0 +1,473 @@
+	}
+
+	dgst := digestMD5{
+		authID:   ourToken.Identifier,
+		passwd:   base64.StdEncoding.EncodeToString(ourToken.Password),
+		hostname: auth.GetServerId(),
+		service:  auth.GetProtocol(),
+	}
+
+	// start handshake
+	var err error
+	if err = binary.Write(conn, binary.BigEndian, uint32(0xDEADBEEF)); err != nil {
+		return nil, err
+	}
+
+	if err = sendSaslMsg(conn, hdfs.DataTransferEncryptorMessageProto_SUCCESS.Enum(), []byte{}, "", false); err != nil {
+		return nil, err
+	}
+
+	msg := &hdfs.DataTransferEncryptorMessageProto{}
+	if err = readPrefixedMessage(conn, msg); err != nil {
+		return nil, err
+	}
+
+	challengeResponse, err := dgst.challengeStep1(msg.Payload)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = sendSaslMsg(conn, hdfs.DataTransferEncryptorMessageProto_SUCCESS.Enum(), []byte(challengeResponse), "", useSecure); err != nil {
+		return nil, err
+	}
+
+	if err = readPrefixedMessage(conn, msg); err != nil {
+		return nil, err
+	}
+
+	// the result of this challenge evaluation should be an empty string
+	if challengeResponse, err = dgst.challengeStep2(msg.Payload); err != nil || len(challengeResponse) != 0 {
+		return nil, err
+	}
+
+	// we're only authenticating, no RPC protection involved, so no wrapping the connection
+	// just return the connection as is, we're done here!
+	if dgst.token.qop == qopAuthentication {
+		return conn, nil
+	}
+
+	kic, kis := generateIntegrityKeyPair(dgst.a1())
+
+	var d digestMD5Handler
+	if dgst.token.qop == qopPrivacy {
+		kcc, kcs := generatePrivacyKeyPair(dgst.a1(), dgst.cipher)
+		d, err = newDigestMD5PrivacyConn(conn, kic[:], kis[:], kcc[:], kcs[:])
+		if err != nil {
+			return nil, err
+		}
+	} else if dgst.token.qop == qopIntegrity {
+		d = newDigestMD5IntegrityConn(conn, kic[:], kis[:])
+	}
+
+	if len(msg.GetCipherOption()) > 0 {
+		cipher := msg.GetCipherOption()[0]
+		var outKey []byte
+
+		decoded, err := d.decode(cipher.InKey)
+		if err != nil {
+			return nil, err
+		}
+		// we reuse the buffer that is returned from decoding, so we need to copy
+		// the decoded output so the next call to decode doesn't blow it away
+		inKey := make([]byte, len(decoded))
+		copy(inKey, decoded)
+
+		if outKey, err = d.decode(cipher.OutKey); err != nil {
+			return nil, err
+		}
+
+		return newAesConn(conn, inKey, outKey, cipher.InIv, cipher.OutIv)
+	}
+	return d, nil
+}
+
+// helper interface for decoding the cipher keys if we need to construct
+// an aes connection
+type digestMD5Handler interface {
+	net.Conn
+	decode(input []byte) ([]byte, error)
+}
+
+// so we don't have to duplicate this for all three types of connection
+// wrappers (md5 integrity, md5 privacy, and aes), have all of them delegate
+// the functions we aren't overriding to the underlying connection object
+// but we don't override the Read and Write here forcing each of the types
+// to have to implement their own in order to meet the net.Conn interface
+type baseConnWrapper struct {
```

Comment:
Let's just duplicate this. Also, why do we need to track the deadline?

Reply:
Because of the tests for the deadline. Since we buffer the reads to make sure that we receive a whole message to decode properly and then store the remaining bytes in the buffer, it doesn't actually go back to the connection until it runs out of data, therefore it won't actually return the timeout unless it has to go back to the connection which the current tests we have don't do. So i had to add this to have it pass the tests. I think if we increase the size of the data being used for the deadline tests it might be able to pass it instead of us tracking the deadline ourselves.

Reply:
That's a good point. It sounds like we should track it, even if we can trick the test. I think that means you only need to track it for `Read`, though, not for `Write`.

Reply:
I've updated to not track the writedeadline and only track the readdeadline.